content/posts/why-meson/index.md

```diff
@@ -0,0 +1,174 @@
+files, functions, compiler or linker arguments.
+
+To some, this may be very similar to what is done by Autotools, however in
+radare2/Rizin case, `configure` is generated by another sh script, `acr`, by
+parsing a `configure.acr` file. `acr` is a tool developed by pancake, the
+author of radare2, and it is an autoconf replacement.
+
+In 2017 pancake started introducing another build system which could be used
+to compile radare2: `meson`. Since then, several other people, including
+myself, has improved this system to compile on several platforms and making
+sure it is (almost) even feature-wise with the ACR/Make build system.
+
+Rizin has chosen to deprecate the use of ACR/Make and switch to `meson` as
+the main build system. We believe this will make the overall build process
+more standard, easy to understand and to integrate with other
+tools/libraries.
+
+# Problems with ACR/Make
+
+There are of course several reasons for this choice,
+so let's first see what *we* believe are the problems of the historical
+approach:
+
+- ACR is essentially a one person project, with mostly only radare2 and other
+radare-related tools using it. This by itself is not a bad thing, but it
+comes with the downside that you find no help or documentation online and if
+you have issues or missing features, you have only to rely on one person who
+understand its internals. Moreover, the features you find are usually just
+the ones used by radare2 project (e.g. not long ago, it was not possible to
+easily check if the compiler supported a particular compilation flag, because
+it was never necessary in radare2).
+- `configure` script needs a sh shell, which makes it hard to use on
+platforms such as Windows. There are of course ways to use it, but they may
+involve installing MinGW or similars, which may not be ideal for Windows
+users who usually work within Visual Studio.
+- Makefiles can be written in a very flexible way and they can be used to
+perform any sort of action, from simply compiling a C file to running scp,
+various scripts and much more. Flexibility shall not be abused though,
+otherwise it may become hard to understand how things are actually done (e.g.
+understanding how `librz_io.so` is compiled involves looking at the Makefile
+in `libr/io`, which includes `config.mk` that setups some variables based on
+other variables defined in the Makefile and then it includes `rules.mk`,
+which uses those variables to actually compile the library. Inside `rules.mk`
+you find, hidden with various environment variables, the commands used to
+build the object files and, then, the library).
+- It is "low-level", which means that the Makefiles specifies the specific
+commands, flags and options that you have to use to actually compile a
+binary, a library or an object file. This provides a lot of power, but it may
+also be overwhelming having to remember to add specific compilation/linking
+flags for compiling a single file.
```

Comment:
Add one more - "speed". Meson allows to use Ninja that outperforms Makefiles N-fold.